---
layout: default
title: Tutorial 4: A practical introduction to multi-omics integration and network analysis
---

#### <img border="0" src="https://s3-us-west-2.amazonaws.com/slack-files2/avatars/2019-09-12/751389607265_d59c0d58846bb2db7123_132.jpg" width="50" height="50"> Tutorial 4: A practical introduction to multi-omics integration and network analysis
[![DOI](https://zenodo.org/badge/172930292.svg)](https://zenodo.org/badge/latestdoi/172930292) <img border="0" src="https://www.svgrepo.com/show/305241/github.svg" width="20" height="20"> [Github repository](https://github.com/NBISweden/workshop_omicsint_ISMBECCB/)    
<br>
This tutorial is a shorter version of the workshop in [Omics Integration and Systems Biology](https://uppsala.instructure.com/courses/52162) of the [National Bioinformatics Infrastructure Sweden](www.nbis.se). Please refer to that workshop for more information. 

<hr>

##### Instructions
Please check the [preparation instructions](./precourse.html). If you are unable to complete all installations you will nevertheless be able to follow all exercises through the respective html files. The [schedule](./schedule.html) contains links to all lectures and notebooks.  

<p style="color:red;font-weight:bold">16/07/2021: Note the updated installation instructions based on Docker instead of Conda.</p>

##### Questions
Please write them on the [HackMD](https://hackmd.io/LI_HCxeRT8-Ty5qjikeFpQ?both)

<hr>

##### Date
Thursday, July 22, 11:00 - 15:00 UTC  
Friday, July 23, 11:00 - 15:00 UTC  

##### Presenters
[Rui Benfeitas](https://nbis.se/about/staff/rui-benfeitas/), Science for Life Laboratory, Stockholm University   
[Nikolay Oskolkov](https://nbis.se/about/staff/nikolay-oskolkov), Science for Life Laboratory, Lund University   
[Ashfaq Ali](https://nbis.se/about/staff/ashfaq-ali/), Science for Life Laboratory, Lund University   
 
##### Summary
Advances in next generation sequencing (NGS) and mass spectrometry have recently allowed us to probe deeper and systematically into different layers of biological information flow. We can now capture snapshots of cellular states at single-cell or tissue levels on genomic, transcriptomic, metabolomic, and proteomic levels, to examine relationships between thousands of features in each of these omics and a given phenotype or disease. However, characterization beyond individual omic levels to understand how multi-omic relationships jointly relate with a given phenotype remains a challenge. How may identify the features with the largest phenotypic impact, and how can we identify patterns among the different layers?

In this tutorial we will introduce several different approaches for integration of multi-omics data including supervised and unsupervised learning and network analyses. We will highlight some of the key issues in dealing with the high multidimensionality that characterizes multi-omic data and techniques to address them. We will also discuss some of the most successful methods for multi-omic data abstraction, and how machine learning approaches can be used in unraveling biological relationships. We will show how biological network analyses can be used to identify patterns within and between omics, and how communities of features may be related with phenotypic data and biologic functions. Finally, we will discuss how meta-analyses and network meta-analyses can be used in analyzing studies from independent experiments.

##### Learning Objectives
1. Identify common issues in integration of highly multidimensional omics data.
2. Identify key methods for data integration through supervised and unsupervised machine learning approaches.
3. Understand how biological network analysis may assist in identifying coordinated patterns between features and associating feature communities with phenomic and biological functions.
4. Hands-on experience in supervised/unsupervised integration and biological network analysis.

##### Audience and level
Aimed at bioinformaticians and computational biologists with experience in analysis of high throughput data and basic statistics knowledge, with R or Python coding experience. Knowledge of machine learning techniques is advantageous. Hands-on sessions will comprise both R and Python coding.

