<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ashfaq Ali" />

<meta name="date" content="2020-09-10" />

<title>Cytoscape_R</title>

<script src="RCircos-HiveR_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="RCircos-HiveR_files/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="RCircos-HiveR_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="RCircos-HiveR_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="RCircos-HiveR_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="RCircos-HiveR_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="RCircos-HiveR_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="RCircos-HiveR_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="RCircos-HiveR_files/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
      a.sourceLine { display: inline-block; line-height: 1.25; }
  a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
  a.sourceLine:empty { height: 1.2em; }
  .sourceCode { overflow: visible; }
  code.sourceCode { white-space: pre; position: relative; }
  div.sourceCode { margin: 1em 0; }
  pre.sourceCode { margin: 0; }
  @media screen {
  div.sourceCode { overflow: auto; }
  }
  @media print {
  code.sourceCode { white-space: pre-wrap; }
  a.sourceLine { text-indent: -1em; padding-left: 1em; }
  }
  pre.numberSource a.sourceLine
    { position: relative; left: -4em; }
  pre.numberSource a.sourceLine::before
    { content: attr(data-line-number);
      position: relative; left: -1em; text-align: right; vertical-align: baseline;
      border: none; pointer-events: all; display: inline-block;
      -webkit-touch-callout: none; -webkit-user-select: none;
      -khtml-user-select: none; -moz-user-select: none;
      -ms-user-select: none; user-select: none;
      padding: 0 4px; width: 4em;
      color: #aaaaaa;
    }
  pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
  div.sourceCode
    { background-color: #f8f8f8; }
  @media screen {
  a.sourceLine::before { text-decoration: underline; }
  }
  code span.al { color: #ef2929; } /* Alert */
  code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
  code span.at { color: #c4a000; } /* Attribute */
  code span.bn { color: #0000cf; } /* BaseN */
  code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
  code span.ch { color: #4e9a06; } /* Char */
  code span.cn { color: #000000; } /* Constant */
  code span.co { color: #8f5902; font-style: italic; } /* Comment */
  code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
  code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
  code span.dt { color: #204a87; } /* DataType */
  code span.dv { color: #0000cf; } /* DecVal */
  code span.er { color: #a40000; font-weight: bold; } /* Error */
  code span.ex { } /* Extension */
  code span.fl { color: #0000cf; } /* Float */
  code span.fu { color: #000000; } /* Function */
  code span.im { } /* Import */
  code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
  code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
  code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
  code span.ot { color: #8f5902; } /* Other */
  code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
  code span.sc { color: #000000; } /* SpecialChar */
  code span.ss { color: #4e9a06; } /* SpecialString */
  code span.st { color: #4e9a06; } /* String */
  code span.va { color: #000000; } /* Variable */
  code span.vs { color: #4e9a06; } /* VerbatimString */
  code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Cytoscape_R</h1>
<h4 class="author">Ashfaq Ali</h4>
<h4 class="date">09/10/2020</h4>

</div>


<!-- rmd lab header -->
<!-- custom fonts -->
<p><link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;subset=latin-ext" rel="stylesheet"></p>
<p><br></p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># suppressMessages(library(dplyr))</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co"># suppressMessages(library(purrr))</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(stringi))</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(tidyr))</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(tidyverse))</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(plyranges))</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(circlize))</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(IRanges))</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(data.table))</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(grid))</a>
<a class="sourceLine" id="cb1-11" data-line-number="11"></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co"># library(HiveR)</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co"># library(rgl)</span></a></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
Output
</div>
</div>
<div id="data-visualization" class="section level1">
<h1><span class="header-section-number">1</span> Data visualization</h1>
<p>Data visualization is part art and part science. The challenge is to get the art right without getting the science wrong and vice versa.</p>
<p>Anscombe’s quartet notwithstanding, and especially for large volumes of data, summary statistics and model estimates should be thought of as tools that we use to deliberately simplify things in a way that lets us see past a cloud of data points shown in a figure. We will not automatically get the right answer to our questions just by looking.</p>
<p>Tufte’s message is sometimes frustrating, but it is consistent:</p>
<p>Graphical excellence is the well-designed presentation of interesting data—a matter of substance, of statistics, and of design … [It] consists of complex ideas communicated with clarity, precision, and efficiency. … [It] is that which gives to the viewer the greatest number of ideas in the shortest time with the least ink in the smallest space … [It] is nearly always multivariate … And graphical excellence requires telling the truth about the data. (Tufte, 1983, p. 51).</p>
<p>To get more insight into data visualization and principles one should follow, please refer to following two books from where I have copied the above text.</p>
<ol style="list-style-type: decimal">
<li>Data Visualization: A practical introduction by Kieran Healy (<a href="https://socviz.co/lookatdata.html" class="uri">https://socviz.co/lookatdata.html</a>)</li>
<li>Fundamentals of Data Visualization by Claus O. Wilke (<a href="https://clauswilke.com/dataviz/" class="uri">https://clauswilke.com/dataviz/</a>)</li>
</ol>
</div>
<div id="circos-plots" class="section level1">
<h1><span class="header-section-number">2</span> Circos Plots</h1>
<p>The circos plot capability to plot genomic data with chromosome coordinates that requires chromosom <em>start</em> and the <em>end</em> site indicating the length of the chromosome. Each chromosom will occupy a <em>sector</em> in the circos plot. Tracks can then be added using the coordinate within each sector (chromosome) provided and relecvant data for each value of the coordinates can be plotte for the respective track. Examples could be genes, transcripts, bases (SNPs, CpG sites) etc, each with start and the end site.</p>
<p>The <em>ciclize</em> package in R and <em>circos</em> algorithm in perl provide easy infrastructure to plot geneomic data.</p>
<p>Circular layout is very useful to represent complicated information.
- it elegantly represents information with long axes or a large amount of categories
- it intuitively shows data with multiple tracks focusing on the same object
- it easily demonstrates relations between elements
- it os aesthetically pleasing way of representing data</p>
<p>Circos plots provide and excellent framework to visulaize multi assay omics data. The circular layout allows us to plot several layers of data based on coordinates. Genomic coordinates are one good example where the whole genome is displayed in the form of coordintes. The genomic coordinates can then be used to stack different layers of data.</p>
<p>A circular layout is composed of sectors and tracks. For data in different categories, they are allocated into different sectors, and for multiple measurements on the same category, they are represented as stacked tracks from outside of the circle to the inside. The intersection of a sector and a track is called a cell (or a grid, a panel), which is the basic unit in a circular layout. It is an imaginary plotting region for drawing data points.</p>
</div>
<div id="principles-of-creating-circos-plot" class="section level1">
<h1><span class="header-section-number">3</span> Principles of creating circos plot</h1>
<p>The routine in making a circos plot works like</p>
<p><strong>initialize layout -&gt; create track -&gt; add graphics -&gt; create track -&gt; add graphics - … -&gt; clear</strong></p>
<div id="a-basic-example-of-circos-plot" class="section level2">
<h2><span class="header-section-number">3.1</span> A basic example of circos plot</h2>
<p>Let us create a small dataset and</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">999</span>)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">n =<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3">df =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">factors =</span> <span class="kw">sample</span>(letters[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>], n, <span class="dt">replace =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">    <span class="dt">x =</span> <span class="kw">rnorm</span>(n), <span class="dt">y =</span> <span class="kw">runif</span>(n), <span class="dt">z=</span><span class="kw">rnorm</span>(n))</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="kw">head</span>(df)</a></code></pre></div>
<pre><code>##   factors           x         y          z
## 1       c -0.33837650 0.2277292 -2.3816263
## 2       d  0.20128773 0.8345422 -0.3889743
## 3       e  0.60271173 0.6507941 -1.2688648
## 4       g -0.22501815 0.8613092 -1.9252602
## 5       a -0.01011359 0.3624672  0.3384133
## 6       f -0.19359724 0.5251656 -1.1476363</code></pre>
</div>
</div>
<div id="initialize-the-circular-layout" class="section level2">
<h2><span class="header-section-number">3.2</span> Initialize the circular layout</h2>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">circos.par</span>(<span class="st">&quot;track.height&quot;</span> =<span class="st"> </span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw">circos.initialize</span>(<span class="dt">factors =</span> df<span class="op">$</span>factors, <span class="dt">x =</span> df<span class="op">$</span>x)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">## create track</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="kw">circos.track</span>(<span class="dt">factors =</span> df<span class="op">$</span>factors, <span class="dt">y =</span> df<span class="op">$</span>y,</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">    <span class="dt">panel.fun =</span> <span class="cf">function</span>(x, y) {</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">        <span class="kw">circos.text</span>(CELL_META<span class="op">$</span>xcenter, CELL_META<span class="op">$</span>cell.ylim[<span class="dv">2</span>] <span class="op">+</span><span class="st"> </span><span class="kw">mm_y</span>(<span class="dv">5</span>), </a>
<a class="sourceLine" id="cb4-10" data-line-number="10">            CELL_META<span class="op">$</span>sector.index)</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">        <span class="kw">circos.axis</span>(<span class="dt">labels.cex =</span> <span class="fl">0.6</span>)</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">})</a>
<a class="sourceLine" id="cb4-13" data-line-number="13"></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">## add graphics</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="kw">circos.par</span>(<span class="st">&quot;track.height&quot;</span> =<span class="st"> </span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="kw">circos.initialize</span>(<span class="dt">factors =</span> df<span class="op">$</span>factors, <span class="dt">x =</span> df<span class="op">$</span>x)</a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="kw">circos.track</span>(<span class="dt">factors =</span> df<span class="op">$</span>factors, <span class="dt">y =</span> df<span class="op">$</span>y,</a>
<a class="sourceLine" id="cb4-21" data-line-number="21">    <span class="dt">panel.fun =</span> <span class="cf">function</span>(x, y) {</a>
<a class="sourceLine" id="cb4-22" data-line-number="22">        <span class="kw">circos.text</span>(CELL_META<span class="op">$</span>xcenter, CELL_META<span class="op">$</span>cell.ylim[<span class="dv">2</span>] <span class="op">+</span><span class="st"> </span><span class="kw">mm_y</span>(<span class="dv">5</span>), </a>
<a class="sourceLine" id="cb4-23" data-line-number="23">            CELL_META<span class="op">$</span>sector.index)</a>
<a class="sourceLine" id="cb4-24" data-line-number="24">        <span class="kw">circos.axis</span>(<span class="dt">labels.cex =</span> <span class="fl">0.6</span>)</a>
<a class="sourceLine" id="cb4-25" data-line-number="25">})</a>
<a class="sourceLine" id="cb4-26" data-line-number="26">col =<span class="st"> </span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;#FF0000&quot;</span>, <span class="st">&quot;#00FF00&quot;</span>), <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb4-27" data-line-number="27"><span class="kw">circos.trackPoints</span>(df<span class="op">$</span>factors, df<span class="op">$</span>x, df<span class="op">$</span>y, <span class="dt">col =</span> col, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb4-28" data-line-number="28"><span class="kw">circos.text</span>(<span class="op">-</span><span class="dv">1</span>, <span class="fl">0.5</span>, <span class="st">&quot;text&quot;</span>, <span class="dt">sector.index =</span> <span class="st">&quot;a&quot;</span>, <span class="dt">track.index =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb4-29" data-line-number="29"></a>
<a class="sourceLine" id="cb4-30" data-line-number="30"></a>
<a class="sourceLine" id="cb4-31" data-line-number="31"><span class="co">## Add more tracks </span></a>
<a class="sourceLine" id="cb4-32" data-line-number="32"></a>
<a class="sourceLine" id="cb4-33" data-line-number="33">bgcol =<span class="st"> </span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;#EFEFEF&quot;</span>, <span class="st">&quot;#CCCCCC&quot;</span>), <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb4-34" data-line-number="34"><span class="kw">circos.trackHist</span>(df<span class="op">$</span>factors, df<span class="op">$</span>z, <span class="dt">bin.size =</span> <span class="fl">0.2</span>, <span class="dt">bg.col =</span> bgcol, <span class="dt">col =</span> <span class="ot">NA</span>)</a>
<a class="sourceLine" id="cb4-35" data-line-number="35"></a>
<a class="sourceLine" id="cb4-36" data-line-number="36"><span class="kw">circos.track</span>(<span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">panel.fun =</span> <span class="cf">function</span>(x, y) {</a>
<a class="sourceLine" id="cb4-37" data-line-number="37">    xlim =<span class="st"> </span>CELL_META<span class="op">$</span>xlim</a>
<a class="sourceLine" id="cb4-38" data-line-number="38">    ylim =<span class="st"> </span>CELL_META<span class="op">$</span>ylim</a>
<a class="sourceLine" id="cb4-39" data-line-number="39">    breaks =<span class="st"> </span><span class="kw">seq</span>(xlim[<span class="dv">1</span>], xlim[<span class="dv">2</span>], <span class="dt">by =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb4-40" data-line-number="40">    n_breaks =<span class="st"> </span><span class="kw">length</span>(breaks)</a>
<a class="sourceLine" id="cb4-41" data-line-number="41">    <span class="kw">circos.rect</span>(breaks[<span class="op">-</span>n_breaks], <span class="kw">rep</span>(ylim[<span class="dv">1</span>], n_breaks <span class="op">-</span><span class="st"> </span><span class="dv">1</span>),</a>
<a class="sourceLine" id="cb4-42" data-line-number="42">                breaks[<span class="op">-</span><span class="dv">1</span>], <span class="kw">rep</span>(ylim[<span class="dv">2</span>], n_breaks <span class="op">-</span><span class="st"> </span><span class="dv">1</span>),</a>
<a class="sourceLine" id="cb4-43" data-line-number="43">                <span class="dt">col =</span> <span class="kw">rand_color</span>(n_breaks), <span class="dt">border =</span> <span class="ot">NA</span>)</a>
<a class="sourceLine" id="cb4-44" data-line-number="44">})</a>
<a class="sourceLine" id="cb4-45" data-line-number="45"></a>
<a class="sourceLine" id="cb4-46" data-line-number="46"></a>
<a class="sourceLine" id="cb4-47" data-line-number="47"><span class="co">## Create links </span></a>
<a class="sourceLine" id="cb4-48" data-line-number="48"></a>
<a class="sourceLine" id="cb4-49" data-line-number="49"><span class="kw">circos.link</span>(<span class="st">&quot;a&quot;</span>, <span class="dv">0</span>, <span class="st">&quot;b&quot;</span>, <span class="dv">0</span>, <span class="dt">h =</span> <span class="fl">0.4</span>)</a>
<a class="sourceLine" id="cb4-50" data-line-number="50"><span class="kw">circos.link</span>(<span class="st">&quot;c&quot;</span>, <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>), <span class="st">&quot;d&quot;</span>, <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>,<span class="fl">0.5</span>), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>,</a>
<a class="sourceLine" id="cb4-51" data-line-number="51">    <span class="dt">border =</span> <span class="st">&quot;blue&quot;</span>, <span class="dt">h =</span> <span class="fl">0.2</span>)</a>
<a class="sourceLine" id="cb4-52" data-line-number="52"><span class="kw">circos.link</span>(<span class="st">&quot;e&quot;</span>, <span class="dv">0</span>, <span class="st">&quot;g&quot;</span>, <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>), <span class="dt">col =</span> <span class="st">&quot;green&quot;</span>, <span class="dt">border =</span> <span class="st">&quot;black&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">lty =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb4-53" data-line-number="53"></a>
<a class="sourceLine" id="cb4-54" data-line-number="54"></a>
<a class="sourceLine" id="cb4-55" data-line-number="55"></a>
<a class="sourceLine" id="cb4-56" data-line-number="56"><span class="kw">circos.clear</span>()</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-5-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /><img src="RCircos-HiveR_files/figure-html/unnamed-chunk-5-2.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
<p>A circular layout is composed of sectors and tracks. For data in different categories, they are allocated into different sectors, and for multiple measurements on the same category, they are represented as stacked tracks from outside of the circle to the inside. The intersection of a sector and a track is called a cell (or a grid, a panel), which is the basic unit in a circular layout. It is an imaginary plotting region for drawing data points.</p>
<p><em>low-level graphic functions that can be used for adding graphics.</em></p>
<ul>
<li>circos.points() : adds points in a cell.</li>
<li>circos.lines() : adds lines in a cell.</li>
<li>circos.segments() : adds segments in a cell.</li>
<li>circos.rect()`: adds rectangles in a cell.</li>
<li>circos.polygon(): adds polygons in a cell.</li>
<li>circos.text(): adds text in a cell.</li>
<li>circos.axis() ands circos.yaxis(): add axis in a cell.</li>
</ul>
<p><em>Following function draws links between two positions in the circle:</em></p>
<p>circos.link()</p>
<p><em>Following functions draw high-level graphics:</em></p>
<p>circos.barplot(): draw barplots.
circos.boxplot(): draw boxplots.
circos.violin(): draws violin plots.
circos.heatmap(): draw circular heatmaps.
circos.raster(): draw raster images.
circos.arrow(): draw circular arrows.</p>
<p><em>Following functions arrange the circular layout.</em></p>
<p>circos.initialize(): allocates sectors on the circle.
circos.track(): creates plotting regions for cells in one single track.
circos.update(): updates an existed cell.
circos.par(): graphic parameters.
circos.info(): prints general parameters of current circular plot.
circos.clear(): resets graphic parameters and internal variables.</p>
</div>
<div id="applications-in-genomicsomics" class="section level2">
<h2><span class="header-section-number">3.3</span> Applications in Genomics/omics</h2>
<p>Let us creat a bed file using the generateRandomBed() function from the “circlize” package.</p>
</div>
<div id="example-for-genomic-data" class="section level2">
<h2><span class="header-section-number">3.4</span> Example for genomic data</h2>
</div>
<div id="circos.genomicpostransformlines" class="section level2">
<h2><span class="header-section-number">3.5</span> circos.genomicPosTransformLines</h2>
<p>For example, there are two sets of regions in a chromosome in which regions in one set regions are quite densely to each other and regions in other set are far from others. Heatmap or text is going to be drawn on the next track. If there is no position transformation, heatmap or text for those dense regions would be overlapped and hard to identify, also ugly to visualize. Thus, a way to transform original positions to new positions would help for the visualization.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">circos.initializeWithIdeogram</span>()</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">123</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">reds10 &lt;-<span class="st"> </span><span class="kw">rand_color</span>(<span class="dv">10</span>, <span class="dt">hue =</span> <span class="st">&quot;red&quot;</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">greens &lt;-<span class="st"> </span><span class="kw">rand_color</span>(<span class="dv">10</span>, <span class="dt">hue =</span> <span class="st">&quot;green&quot;</span>)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">blues &lt;-<span class="st"> </span><span class="kw">rand_color</span>(<span class="dv">10</span>, <span class="dt">hue =</span> <span class="st">&quot;blue&quot;</span>)</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#circos.genomicPosTransformLines(bed, posTransform = posTransform.default, horizontalLine = &quot;top&quot;, col =reds10[1], lwd = 2)</span></a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-6-1.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
<div id="add-heatmap-to-the-genomic-regions" class="section level2">
<h2><span class="header-section-number">3.6</span> Add heatmap to the genomic regions</h2>
<p>The main idea for plotting data for circos plots is that one passes the data circos.genomicTrackPlotRegion() function together with a panel.function. Within the panel function argument, the relevant plotting function can be supplied togther with the relevant data. The relevant data maybe a bed file as a data frame where first three columns the file refer to the coordinates while column 4 and onwards refer to the data values to be plotted.</p>
<p>Here we want to plot heatmap which is done by circos.genomicRect() function that expects</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">circos.initializeWithIdeogram</span>()</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">400</span>, <span class="dt">nc =</span> <span class="dv">8</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">om =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;track.margin&quot;</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">oc =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;cell.padding&quot;</span>)</a>
<a class="sourceLine" id="cb6-5" data-line-number="5"></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="kw">circos.par</span>(<span class="dt">track.margin =</span> <span class="kw">c</span>(om[<span class="dv">1</span>], <span class="dv">0</span>), <span class="dt">cell.padding =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb6-7" data-line-number="7"></a>
<a class="sourceLine" id="cb6-8" data-line-number="8">f =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="dt">breaks =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))</a>
<a class="sourceLine" id="cb6-9" data-line-number="9"></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb6-11" data-line-number="11">    <span class="kw">circos.genomicRect</span>(<span class="dt">region =</span> coord, <span class="dt">value =</span> vals, <span class="dt">col =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), </a>
<a class="sourceLine" id="cb6-12" data-line-number="12">        <span class="dt">border =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), <span class="dt">lwd =</span> <span class="fl">0.1</span>, <span class="dt">posTransform =</span> posTransform.default, ...)</a>
<a class="sourceLine" id="cb6-13" data-line-number="13">}, <span class="dt">bg.border =</span> <span class="ot">NA</span>, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb6-14" data-line-number="14"></a>
<a class="sourceLine" id="cb6-15" data-line-number="15"></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="kw">circos.par</span>(<span class="dt">track.margin =</span> om, <span class="dt">cell.padding =</span> oc)</a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="kw">circos.clear</span>()</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-7-1.svg" width="1440" style="display: block; margin: auto auto auto 0;" />
</div>
<p>Another way of plotting heatmaps is using circos.genomicHeatmap() function.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">circos.initializeWithIdeogram</span>()</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">400</span>, <span class="dt">nc =</span> <span class="dv">8</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">om =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;track.margin&quot;</span>)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">oc =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;cell.padding&quot;</span>)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5"></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="kw">circos.par</span>(<span class="dt">track.margin =</span> <span class="kw">c</span>(om[<span class="dv">1</span>], <span class="dv">0</span>), <span class="dt">cell.padding =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb7-7" data-line-number="7"></a>
<a class="sourceLine" id="cb7-8" data-line-number="8">f =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="dt">breaks =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))</a>
<a class="sourceLine" id="cb7-9" data-line-number="9"></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb7-11" data-line-number="11">    <span class="kw">circos.genomicRect</span>(<span class="dt">region =</span> coord, <span class="dt">value =</span> vals, <span class="dt">col =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), </a>
<a class="sourceLine" id="cb7-12" data-line-number="12">        <span class="dt">border =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), <span class="dt">lwd =</span> <span class="fl">0.1</span>, <span class="dt">posTransform =</span> posTransform.default, ...)</a>
<a class="sourceLine" id="cb7-13" data-line-number="13">}, <span class="dt">bg.border =</span> <span class="ot">NA</span>, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb7-14" data-line-number="14"></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"></a>
<a class="sourceLine" id="cb7-16" data-line-number="16"><span class="kw">circos.clear</span>()</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-8-1.svg" width="1440" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
<div id="points" class="section level2">
<h2><span class="header-section-number">3.7</span> Points</h2>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">circos.initializeWithIdeogram</span>()</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">om =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;track.margin&quot;</span>)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4">oc =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;cell.padding&quot;</span>)</a>
<a class="sourceLine" id="cb8-5" data-line-number="5"></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="kw">circos.par</span>(<span class="dt">track.margin =</span> <span class="kw">c</span>(om[<span class="dv">1</span>], <span class="dv">0</span>), <span class="dt">cell.padding =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb8-7" data-line-number="7"></a>
<a class="sourceLine" id="cb8-8" data-line-number="8">f =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="dt">breaks =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))</a>
<a class="sourceLine" id="cb8-9" data-line-number="9"></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb8-11" data-line-number="11">    <span class="kw">circos.genomicRect</span>(<span class="dt">region =</span> coord, <span class="dt">value =</span> vals, <span class="dt">col =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), </a>
<a class="sourceLine" id="cb8-12" data-line-number="12">        <span class="dt">border =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), <span class="dt">lwd =</span> <span class="fl">0.1</span>, <span class="dt">posTransform =</span> posTransform.default, ...)</a>
<a class="sourceLine" id="cb8-13" data-line-number="13">}, <span class="dt">bg.border =</span> <span class="ot">NA</span>, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb8-14" data-line-number="14"></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"></a>
<a class="sourceLine" id="cb8-16" data-line-number="16">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">900</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">runif</span>(k)<span class="op">*</span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), k, <span class="dt">replace =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb8-17" data-line-number="17"></a>
<a class="sourceLine" id="cb8-18" data-line-number="18"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb8-19" data-line-number="19">    col =<span class="st"> </span><span class="kw">ifelse</span>(vals[[<span class="dv">1</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>)</a>
<a class="sourceLine" id="cb8-20" data-line-number="20">    <span class="kw">circos.genomicPoints</span>(<span class="dt">region=</span>coord, <span class="dt">value=</span>vals, <span class="dt">col =</span> col, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">pch =</span> <span class="dv">16</span>)</a>
<a class="sourceLine" id="cb8-21" data-line-number="21">    cell.xlim =<span class="st"> </span><span class="kw">get.cell.meta.data</span>(<span class="st">&quot;cell.xlim&quot;</span>)</a>
<a class="sourceLine" id="cb8-22" data-line-number="22">    <span class="cf">for</span>(h <span class="cf">in</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="fl">-0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)) {</a>
<a class="sourceLine" id="cb8-23" data-line-number="23">        <span class="kw">circos.lines</span>(cell.xlim, <span class="kw">c</span>(h, h), <span class="dt">col =</span> <span class="st">&quot;#00000040&quot;</span>)</a>
<a class="sourceLine" id="cb8-24" data-line-number="24">    }</a>
<a class="sourceLine" id="cb8-25" data-line-number="25">}, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb8-26" data-line-number="26"><span class="kw">circos.clear</span>()</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-9-1.svg" width="1440" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
<div id="add-complex-tracks" class="section level2">
<h2><span class="header-section-number">3.8</span> Add complex tracks</h2>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">circos.initializeWithIdeogram</span>()</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">om =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;track.margin&quot;</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">oc =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;cell.padding&quot;</span>)</a>
<a class="sourceLine" id="cb9-5" data-line-number="5"></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="kw">circos.par</span>(<span class="dt">track.margin =</span> <span class="kw">c</span>(om[<span class="dv">1</span>], <span class="dv">0</span>), <span class="dt">cell.padding =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb9-7" data-line-number="7"></a>
<a class="sourceLine" id="cb9-8" data-line-number="8">f =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="dt">breaks =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))</a>
<a class="sourceLine" id="cb9-9" data-line-number="9">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">400</span>, <span class="dt">nc =</span> <span class="dv">8</span>)</a>
<a class="sourceLine" id="cb9-10" data-line-number="10"></a>
<a class="sourceLine" id="cb9-11" data-line-number="11"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb9-12" data-line-number="12">    <span class="kw">circos.genomicRect</span>(<span class="dt">region =</span> coord, <span class="dt">value =</span> vals, <span class="dt">col =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), </a>
<a class="sourceLine" id="cb9-13" data-line-number="13">        <span class="dt">border =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), <span class="dt">lwd =</span> <span class="fl">0.1</span>, <span class="dt">posTransform =</span> posTransform.default, ...)</a>
<a class="sourceLine" id="cb9-14" data-line-number="14">}, <span class="dt">bg.border =</span> <span class="ot">NA</span>, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb9-15" data-line-number="15"></a>
<a class="sourceLine" id="cb9-16" data-line-number="16"></a>
<a class="sourceLine" id="cb9-17" data-line-number="17"></a>
<a class="sourceLine" id="cb9-18" data-line-number="18">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">900</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">runif</span>(k)<span class="op">*</span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), k, <span class="dt">replace =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb9-19" data-line-number="19"></a>
<a class="sourceLine" id="cb9-20" data-line-number="20"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb9-21" data-line-number="21">    col =<span class="st"> </span><span class="kw">ifelse</span>(vals[[<span class="dv">1</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>)</a>
<a class="sourceLine" id="cb9-22" data-line-number="22">    <span class="kw">circos.genomicPoints</span>(<span class="dt">region=</span>coord, <span class="dt">value=</span>vals, <span class="dt">col =</span> col, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">pch =</span> <span class="dv">16</span>)</a>
<a class="sourceLine" id="cb9-23" data-line-number="23">    cell.xlim =<span class="st"> </span><span class="kw">get.cell.meta.data</span>(<span class="st">&quot;cell.xlim&quot;</span>)</a>
<a class="sourceLine" id="cb9-24" data-line-number="24">    <span class="cf">for</span>(h <span class="cf">in</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="fl">-0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)) {</a>
<a class="sourceLine" id="cb9-25" data-line-number="25">        <span class="kw">circos.lines</span>(cell.xlim, <span class="kw">c</span>(h, h), <span class="dt">col =</span> <span class="st">&quot;#00000040&quot;</span>)</a>
<a class="sourceLine" id="cb9-26" data-line-number="26">    }</a>
<a class="sourceLine" id="cb9-27" data-line-number="27">}, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb9-28" data-line-number="28"></a>
<a class="sourceLine" id="cb9-29" data-line-number="29"></a>
<a class="sourceLine" id="cb9-30" data-line-number="30">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">900</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">rnorm</span>(k, <span class="dv">0</span>, <span class="dv">50</span>))</a>
<a class="sourceLine" id="cb9-31" data-line-number="31"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {</a>
<a class="sourceLine" id="cb9-32" data-line-number="32">    x =<span class="st"> </span>(region[[<span class="dv">2</span>]] <span class="op">+</span><span class="st"> </span>region[[<span class="dv">1</span>]]) <span class="op">/</span><span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb9-33" data-line-number="33">    y =<span class="st"> </span>value[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb9-34" data-line-number="34">    loess.fit =<span class="st"> </span><span class="kw">loess</span>(y <span class="op">~</span><span class="st"> </span>x)</a>
<a class="sourceLine" id="cb9-35" data-line-number="35">    loess.predict =<span class="st"> </span><span class="kw">predict</span>(loess.fit, x, <span class="dt">se =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb9-36" data-line-number="36">    d1 =<span class="st"> </span><span class="kw">c</span>(x, <span class="kw">rev</span>(x))</a>
<a class="sourceLine" id="cb9-37" data-line-number="37">    d2 =<span class="st"> </span><span class="kw">c</span>(loess.predict<span class="op">$</span>fit <span class="op">+</span><span class="st"> </span>loess.predict<span class="op">$</span>se.fit, <span class="kw">rev</span>(loess.predict<span class="op">$</span>fit <span class="op">-</span><span class="st"> </span>loess.predict<span class="op">$</span>se.fit))</a>
<a class="sourceLine" id="cb9-38" data-line-number="38">    <span class="kw">circos.polygon</span>(d1, d2, <span class="dt">col =</span> <span class="st">&quot;#CCCCCC&quot;</span>, <span class="dt">border =</span> <span class="ot">NA</span>)</a>
<a class="sourceLine" id="cb9-39" data-line-number="39">    <span class="kw">circos.points</span>(x, y, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb9-40" data-line-number="40">    <span class="kw">circos.lines</span>(x, loess.predict<span class="op">$</span>fit)</a>
<a class="sourceLine" id="cb9-41" data-line-number="41">}, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb9-42" data-line-number="42"></a>
<a class="sourceLine" id="cb9-43" data-line-number="43"><span class="kw">circos.clear</span>()</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-10-1.svg" width="1440" style="display: block; margin: auto auto auto 0;" />
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">circos.initializeWithIdeogram</span>()</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">400</span>, <span class="dt">nc =</span> <span class="dv">8</span>)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"></a>
<a class="sourceLine" id="cb10-4" data-line-number="4">om =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;track.margin&quot;</span>)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">oc =<span class="st"> </span><span class="kw">circos.par</span>(<span class="st">&quot;cell.padding&quot;</span>)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6"></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="kw">circos.par</span>(<span class="dt">track.margin =</span> <span class="kw">c</span>(om[<span class="dv">1</span>], <span class="dv">0</span>), <span class="dt">cell.padding =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb10-8" data-line-number="8"></a>
<a class="sourceLine" id="cb10-9" data-line-number="9">f =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="dt">breaks =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))</a>
<a class="sourceLine" id="cb10-10" data-line-number="10"></a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb10-12" data-line-number="12">    <span class="kw">circos.genomicRect</span>(<span class="dt">region =</span> coord, <span class="dt">value =</span> vals, <span class="dt">col =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), </a>
<a class="sourceLine" id="cb10-13" data-line-number="13">        <span class="dt">border =</span> <span class="kw">f</span>(vals[[<span class="dv">1</span>]]), <span class="dt">lwd =</span> <span class="fl">0.1</span>, <span class="dt">posTransform =</span> posTransform.default, ...)</a>
<a class="sourceLine" id="cb10-14" data-line-number="14">}, <span class="dt">bg.border =</span> <span class="ot">NA</span>, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb10-15" data-line-number="15"></a>
<a class="sourceLine" id="cb10-16" data-line-number="16"></a>
<a class="sourceLine" id="cb10-17" data-line-number="17"></a>
<a class="sourceLine" id="cb10-18" data-line-number="18">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">900</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">runif</span>(k)<span class="op">*</span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), k, <span class="dt">replace =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb10-19" data-line-number="19"></a>
<a class="sourceLine" id="cb10-20" data-line-number="20"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">panel.fun =</span> <span class="cf">function</span>(coord, vals, ...) {</a>
<a class="sourceLine" id="cb10-21" data-line-number="21">    col =<span class="st"> </span><span class="kw">ifelse</span>(vals[[<span class="dv">1</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>)</a>
<a class="sourceLine" id="cb10-22" data-line-number="22">    <span class="kw">circos.genomicPoints</span>(<span class="dt">region=</span>coord, <span class="dt">value=</span>vals, <span class="dt">col =</span> col, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">pch =</span> <span class="dv">16</span>)</a>
<a class="sourceLine" id="cb10-23" data-line-number="23">    cell.xlim =<span class="st"> </span><span class="kw">get.cell.meta.data</span>(<span class="st">&quot;cell.xlim&quot;</span>)</a>
<a class="sourceLine" id="cb10-24" data-line-number="24">    <span class="cf">for</span>(h <span class="cf">in</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="fl">-0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)) {</a>
<a class="sourceLine" id="cb10-25" data-line-number="25">        <span class="kw">circos.lines</span>(cell.xlim, <span class="kw">c</span>(h, h), <span class="dt">col =</span> <span class="st">&quot;#00000040&quot;</span>)</a>
<a class="sourceLine" id="cb10-26" data-line-number="26">    }</a>
<a class="sourceLine" id="cb10-27" data-line-number="27">}, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb10-28" data-line-number="28"></a>
<a class="sourceLine" id="cb10-29" data-line-number="29"></a>
<a class="sourceLine" id="cb10-30" data-line-number="30">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">900</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">rnorm</span>(k, <span class="dv">0</span>, <span class="dv">50</span>))</a>
<a class="sourceLine" id="cb10-31" data-line-number="31"><span class="kw">circos.genomicTrackPlotRegion</span>(bed, <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {</a>
<a class="sourceLine" id="cb10-32" data-line-number="32">    x =<span class="st"> </span>(region[[<span class="dv">2</span>]] <span class="op">+</span><span class="st"> </span>region[[<span class="dv">1</span>]]) <span class="op">/</span><span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb10-33" data-line-number="33">    y =<span class="st"> </span>value[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb10-34" data-line-number="34">    loess.fit =<span class="st"> </span><span class="kw">loess</span>(y <span class="op">~</span><span class="st"> </span>x)</a>
<a class="sourceLine" id="cb10-35" data-line-number="35">    loess.predict =<span class="st"> </span><span class="kw">predict</span>(loess.fit, x, <span class="dt">se =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb10-36" data-line-number="36">    d1 =<span class="st"> </span><span class="kw">c</span>(x, <span class="kw">rev</span>(x))</a>
<a class="sourceLine" id="cb10-37" data-line-number="37">    d2 =<span class="st"> </span><span class="kw">c</span>(loess.predict<span class="op">$</span>fit <span class="op">+</span><span class="st"> </span>loess.predict<span class="op">$</span>se.fit, <span class="kw">rev</span>(loess.predict<span class="op">$</span>fit <span class="op">-</span><span class="st"> </span>loess.predict<span class="op">$</span>se.fit))</a>
<a class="sourceLine" id="cb10-38" data-line-number="38">    <span class="kw">circos.polygon</span>(d1, d2, <span class="dt">col =</span> <span class="st">&quot;#CCCCCC&quot;</span>, <span class="dt">border =</span> <span class="ot">NA</span>)</a>
<a class="sourceLine" id="cb10-39" data-line-number="39">    <span class="kw">circos.points</span>(x, y, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb10-40" data-line-number="40">    <span class="kw">circos.lines</span>(x, loess.predict<span class="op">$</span>fit)</a>
<a class="sourceLine" id="cb10-41" data-line-number="41">}, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb10-42" data-line-number="42"></a>
<a class="sourceLine" id="cb10-43" data-line-number="43">bed_list =<span class="st"> </span><span class="kw">list</span>(<span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">900</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">runif</span>(k)),</a>
<a class="sourceLine" id="cb10-44" data-line-number="44">                <span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">900</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">runif</span>(k)))</a>
<a class="sourceLine" id="cb10-45" data-line-number="45">col =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#FF000040&quot;</span>, <span class="st">&quot;#0000FF40&quot;</span>)</a>
<a class="sourceLine" id="cb10-46" data-line-number="46"><span class="kw">circos.genomicTrackPlotRegion</span>(bed_list, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {</a>
<a class="sourceLine" id="cb10-47" data-line-number="47">    i =<span class="st"> </span><span class="kw">getI</span>(...)</a>
<a class="sourceLine" id="cb10-48" data-line-number="48">    <span class="cf">if</span>(i <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb10-49" data-line-number="49">        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">area =</span> <span class="ot">TRUE</span>, <span class="dt">baseline =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">border =</span> <span class="ot">NA</span>, ...)</a>
<a class="sourceLine" id="cb10-50" data-line-number="50">    } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb10-51" data-line-number="51">        <span class="kw">circos.genomicLines</span>(region, <span class="op">-</span>value, <span class="dt">area =</span> <span class="ot">TRUE</span>, <span class="dt">baseline =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span>, <span class="dt">border =</span> <span class="ot">NA</span>, ...)</a>
<a class="sourceLine" id="cb10-52" data-line-number="52">    }</a>
<a class="sourceLine" id="cb10-53" data-line-number="53">}, <span class="dt">track.height =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb10-54" data-line-number="54"></a>
<a class="sourceLine" id="cb10-55" data-line-number="55"><span class="co">## Add links</span></a>
<a class="sourceLine" id="cb10-56" data-line-number="56">region1 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">1500</span>); region1 =<span class="st"> </span>region1[<span class="kw">sample</span>(<span class="kw">nrow</span>(region1), <span class="dv">30</span>), ]</a>
<a class="sourceLine" id="cb10-57" data-line-number="57">region2 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">1500</span>); region2 =<span class="st"> </span>region2[<span class="kw">sample</span>(<span class="kw">nrow</span>(region2), <span class="dv">30</span>), ]</a>
<a class="sourceLine" id="cb10-58" data-line-number="58"><span class="kw">circos.genomicLink</span>(region1, region2, <span class="dt">col =</span> <span class="kw">sample</span>(<span class="dv">30</span>, <span class="dv">30</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb10-59" data-line-number="59"></a>
<a class="sourceLine" id="cb10-60" data-line-number="60"></a>
<a class="sourceLine" id="cb10-61" data-line-number="61"><span class="co">## Highlight chromosomes, tracks and indexes</span></a>
<a class="sourceLine" id="cb10-62" data-line-number="62"></a>
<a class="sourceLine" id="cb10-63" data-line-number="63"></a>
<a class="sourceLine" id="cb10-64" data-line-number="64"><span class="kw">highlight.chromosome</span>(<span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>, <span class="st">&quot;chr12&quot;</span>))</a>
<a class="sourceLine" id="cb10-65" data-line-number="65"><span class="kw">highlight.sector</span>(<span class="kw">c</span>(<span class="st">&quot;chr3&quot;</span>, <span class="st">&quot;chr1&quot;</span>), <span class="dt">track.index =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</a>
<a class="sourceLine" id="cb10-66" data-line-number="66"><span class="kw">circos.clear</span>()</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-11-1.svg" width="1440" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
</div>
<div id="hiveplot" class="section level1">
<h1><span class="header-section-number">4</span> Hiveplot</h1>
<p>Hive plots are an alternative way of visualizing the networks where different parameters and attributes of networks can be plotted on multiple linear axis. The attibutes of nodes and edges are very similar to a network except that instead of trying to intepret intesities of colors, degree of nodes etc. in a network, one can compare network paramaters on a linear scale. An advantage of hive plots for networks is that the subjective observations of nodes and edges of network layout can be obejectively compared.</p>
<p>It all sounds wordy but once you see your results in prtactice, it is much easier to understand.
The interpretation of values mapped to the axis can</p>
<ol style="list-style-type: decimal">
<li>Create Sif file</li>
<li>Create Gsc file
Here I quote HiveR paper for intuition behind hive plots.</li>
</ol>
<p>“HiveR was inspired by the concept of hive plots as developed by Martin Krzy- winski at the Genome Sciences Center (www.hiveplot.com). Hive plots are a reaction to “hairball” style networks in which the layout of the network is arbi- trary and hypersensitive to even small changes in the underlying network. Hive plots are particularly well-suited for comparing networks, as well as for the discovery of emergent properties of networks.
The key innovation in a hive plot, compared to other means of graphically displaying network structure, is how node information is handled. In a hive plot, there is a node coordinate system consisting of two parts. First, nodes are assigned to axes based upon qualitative or quantitative characteristics of the the node, for instance membership in a certain category. As will be discussed later, this assignment process is key to constructing a hive plot. Second, the position of the node along the axis, the radius, is based upon some quantitative characteristic of the node. Edges are handled in a fairly standard way, but may be colored or have a width or weight which encodes an interesting value.
In creating a hive plot, one maps network parameters to the plot, and thus the process can be readily tuned to meet one’s needs. The mappable parame- ters are listed in Table 1, and the mapping is limited only by one’s creativity and the particular knowledge domain. Thus ecologists have their own measures of food webs, social network analysts have various measures describing intercon- nectedness etc. An essential point is that mapping network parameters in this way results in a reproducible plot.&quot;</p>
<p><strong>For this session you may have to install HiveR package to your current conda environment manually as it in currently present in your environment.</strong></p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co">#install.packages(&quot;HiveR&quot;)</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw">suppressMessages</span>(<span class="kw">suppressWarnings</span>(<span class="kw">library</span>(HiveR)))</a></code></pre></div>
</div>
<div id="hiver-functionality" class="section level2">
<h2><span class="header-section-number">4.1</span> HiveR functionality</h2>
<p>To explore more about the hive plot let us look at the package vignette.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">browseVignettes</span>(<span class="st">&quot;HiveR&quot;</span>)</a></code></pre></div>
</div>
<p>HiveR Vignette provides an overview of the package design and functions. Just like Circlize package rand hive plot data can be created using ranHiveData functions.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">ranHive &lt;-<span class="st"> </span><span class="kw">ranHiveData</span>(<span class="dt">nx=</span><span class="dv">4</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw">class</span>(ranHive)</a></code></pre></div>
<pre><code>## [1] &quot;HivePlotData&quot;</code></pre>
</div>
<p>As we can see that ranHive data belongs to “HivePlotData” class and we can explore with sort of data and attributes it is required to have.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw">help</span>(<span class="st">&quot;HivePlotData&quot;</span>)</a></code></pre></div>
</div>
<p>Lets explore the data we created using ranHiveData generator.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw">str</span>(ranHive)</a></code></pre></div>
<pre><code>## List of 5
##  $ nodes    :&#39;data.frame&#39;:   48 obs. of  6 variables:
##   ..$ id    : int [1:48] 3 4 5 6 8 9 10 11 12 13 ...
##   ..$ lab   : chr [1:48] &quot;strma&quot; &quot;pdqfb&quot; &quot;dfgyl&quot; &quot;dggnd&quot; ...
##   ..$ axis  : int [1:48] 4 3 3 1 2 3 4 2 4 4 ...
##   ..$ radius: num [1:48] 16 26 8 19 52 41 25 36 62 65 ...
##   ..$ size  : num [1:48] 1 1.5 1 1.5 1.5 0.5 0.5 1.5 1.5 0.5 ...
##   ..$ color : chr [1:48] &quot;#4DAF4A&quot; &quot;#984EA3&quot; &quot;#E41A1C&quot; &quot;#984EA3&quot; ...
##  $ edges    :&#39;data.frame&#39;:   60 obs. of  4 variables:
##   ..$ id1   : int [1:60] 36 25 39 6 24 57 39 54 39 14 ...
##   ..$ id2   : int [1:60] 44 11 45 17 42 8 42 45 44 49 ...
##   ..$ weight: num [1:60] 2 3 1 3 1 1 3 1 2 2 ...
##   ..$ color : chr [1:60] &quot;#E41A1C&quot; &quot;#377EB8&quot; &quot;#4DAF4A&quot; &quot;#FF7F00&quot; ...
##  $ type     : chr &quot;2D&quot;
##  $ desc     : chr &quot;4 axes -- 48 nodes -- 60 edges&quot;
##  $ axis.cols: chr [1:4] &quot;#E41A1C&quot; &quot;#377EB8&quot; &quot;#4DAF4A&quot; &quot;#984EA3&quot;
##  - attr(*, &quot;class&quot;)= chr &quot;HivePlotData&quot;</code></pre>
</div>
</div>
<div id="plot-the-generated-dat" class="section level2">
<h2><span class="header-section-number">4.2</span> Plot the generated dat</h2>
Let us make a plot to see an example of a hive plot.
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw">plotHive</span>(ranHive)</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-17-1.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
<div id="data-import" class="section level2">
<h2><span class="header-section-number">4.3</span> Data import</h2>
<p>There are two ways of importing data into “HivePlotData” class. The HiveR package “dot2HPD” provides an option to import networks in the dot file format. One can read more about the file format here (<a href="https://secure.wikimedia.org/wikipedia/en/wiki/DOT_language" class="uri">https://secure.wikimedia.org/wikipedia/en/wiki/DOT_language</a>).</p>
<p>A second way of creating “HivePlotData” is from your own netwrok files. Here we will first use - “E.Coli” RegulonDB provided in the HiveR package by using built in functions
- Use Piano package’s output to create Hive plot from your own data by converting it to “HivePlotData” class.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"> EC1 &lt;-<span class="st"> </span><span class="kw">dot2HPD</span>(<span class="dt">file =</span> <span class="st">&quot;Circos_HiveR/hiveData/E_coli/E_coli_TF.dot&quot;</span>, <span class="dt">node.inst =</span> <span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="dt">edge.inst =</span> <span class="st">&quot;Circos_HiveR/hiveData/E_coli/EdgeInst_TF.csv&quot;</span>, <span class="dt">desc =</span> <span class="st">&quot;E coli gene regulatory network (RegulonDB)&quot;</span>, <span class="dt">axis.cols =</span> <span class="kw">rep</span>(<span class="st">&quot;grey&quot;</span>, <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="kw">sumHPD</span>(EC1)</a></code></pre></div>
<pre><code>##  E coli gene regulatory network (RegulonDB)
##  This hive plot data set contains 1597 nodes on 1 axes and 3893 edges.
##  It is a  2D data set.
## 
##      Axis 1 has 1597 nodes spanning radii from 1 to 1 
## 
##      Axes 1 and 1 share 3893 edges</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">EC2 &lt;-<span class="st"> </span><span class="kw">mineHPD</span>(EC1, <span class="dt">option =</span> <span class="st">&quot;rad &lt;- tot.edge.count&quot;</span>)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">sumHPD</span>(EC2)</a></code></pre></div>
<pre><code>##  E coli gene regulatory network (RegulonDB)
##  This hive plot data set contains 1597 nodes on 1 axes and 3893 edges.
##  It is a  2D data set.
## 
##      Axis 1 has 1597 nodes spanning radii from 1 to 434 
## 
##      Axes 1 and 1 share 3893 edges</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">EC3 &lt;-<span class="st"> </span><span class="kw">mineHPD</span>(EC2, <span class="dt">option =</span> <span class="st">&quot;axis &lt;- source.man.sink&quot;</span>)</a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="kw">sumHPD</span>(EC3)</a></code></pre></div>
<pre><code>##  E coli gene regulatory network (RegulonDB)
##  This hive plot data set contains 1597 nodes on 3 axes and 3893 edges.
##  It is a  2D data set.
## 
##      Axis 1 has 45 nodes spanning radii from 1 to 83 
##      Axis 2 has 1416 nodes spanning radii from 1 to 11 
##      Axis 3 has 136 nodes spanning radii from 2 to 434 
## 
##      Axes 1 and 2 share 400 edges
##      Axes 1 and 3 share 21 edges
##      Axes 3 and 2 share 3158 edges
##      Axes 3 and 3 share 314 edges</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">EC4 &lt;-<span class="st"> </span><span class="kw">mineHPD</span>(EC3, <span class="dt">option =</span> <span class="st">&quot;remove zero edge&quot;</span>)</a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="kw">sumHPD</span>(EC4)</a></code></pre></div>
<pre><code>## 
##   113 edge(s) that start and end on the same node were removed
## 
##   22 virtual self-edge(s) were removed
##  E coli gene regulatory network (RegulonDB)
##  This hive plot data set contains 1597 nodes on 3 axes and 3768 edges.
##  It is a  2D data set.
## 
##      Axis 1 has 45 nodes spanning radii from 1 to 83 
##      Axis 2 has 1416 nodes spanning radii from 1 to 11 
##      Axis 3 has 136 nodes spanning radii from 2 to 434 
## 
##      Axes 1 and 2 share 400 edges
##      Axes 1 and 3 share 21 edges
##      Axes 3 and 2 share 3158 edges
##      Axes 3 and 3 share 189 edges</code></pre>
</div>
</div>
<div id="work-with-edges" class="section level2">
<h2><span class="header-section-number">4.4</span> Work with edges</h2>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1">edges &lt;-<span class="st"> </span>EC4<span class="op">$</span>edges</a>
<a class="sourceLine" id="cb27-2" data-line-number="2">edgesR &lt;-<span class="st"> </span><span class="kw">subset</span>(edges, color <span class="op">==</span><span class="st"> &#39;red&#39;</span>)</a>
<a class="sourceLine" id="cb27-3" data-line-number="3">edgesG &lt;-<span class="st"> </span><span class="kw">subset</span>(edges, color <span class="op">==</span><span class="st"> &#39;green&#39;</span>)</a>
<a class="sourceLine" id="cb27-4" data-line-number="4">edgesO &lt;-<span class="st"> </span><span class="kw">subset</span>(edges, color <span class="op">==</span><span class="st"> &#39;orange&#39;</span>)</a>
<a class="sourceLine" id="cb27-5" data-line-number="5"></a>
<a class="sourceLine" id="cb27-6" data-line-number="6">edges &lt;-<span class="st"> </span><span class="kw">rbind</span>(edgesO, edgesG, edgesR)</a>
<a class="sourceLine" id="cb27-7" data-line-number="7">EC4<span class="op">$</span>edges &lt;-<span class="st"> </span>edges</a>
<a class="sourceLine" id="cb27-8" data-line-number="8"></a>
<a class="sourceLine" id="cb27-9" data-line-number="9">EC4<span class="op">$</span>edges<span class="op">$</span>weight &lt;-<span class="st"> </span><span class="fl">0.5</span></a></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw">plotHive</span>(EC4, <span class="dt">dr.nodes =</span> <span class="ot">FALSE</span>, <span class="dt">ch =</span> <span class="dv">20</span>,</a>
<a class="sourceLine" id="cb28-2" data-line-number="2"><span class="dt">axLabs =</span> <span class="kw">c</span>(<span class="st">&quot;source&quot;</span>, <span class="st">&quot;sink&quot;</span>, <span class="st">&quot;manager&quot;</span>),</a>
<a class="sourceLine" id="cb28-3" data-line-number="3"><span class="dt">axLab.pos =</span> <span class="kw">c</span>(<span class="dv">40</span>, <span class="dv">75</span>, <span class="dv">35</span>),</a>
<a class="sourceLine" id="cb28-4" data-line-number="4"><span class="dt">axLab.gpar =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">6</span>, <span class="dt">col =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb28-5" data-line-number="5"><span class="dt">arrow =</span> <span class="kw">c</span>(<span class="st">&quot;degree&quot;</span>, <span class="dv">150</span>, <span class="dv">100</span>, <span class="dv">180</span>, <span class="dv">70</span>), <span class="dt">np =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb28-6" data-line-number="6"><span class="kw">grid.text</span>(<span class="st">&quot;native units&quot;</span>, <span class="dt">x =</span> <span class="fl">0.5</span>, <span class="dt">y =</span> <span class="fl">0.05</span>, <span class="dt">default.units =</span> <span class="st">&quot;npc&quot;</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>, <span class="dt">col =</span> <span class="st">&quot;white&quot;</span>))</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-23-1.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw">plotHive</span>(EC4, <span class="dt">dr.nodes =</span> <span class="ot">FALSE</span>, <span class="dt">method =</span> <span class="st">&quot;rank&quot;</span>, <span class="dt">ch =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="co">#axLabs = c(&quot;source&quot;, &quot;sink&quot;, &quot;manager&quot;),</span></a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="co">#axLab.pos = c(100, 125, 180),</span></a>
<a class="sourceLine" id="cb29-4" data-line-number="4"><span class="co">#axLab.gpar = gpar(fontsize = 10, col = &quot;white&quot;),</span></a>
<a class="sourceLine" id="cb29-5" data-line-number="5"><span class="dt">np =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb29-6" data-line-number="6"><span class="kw">grid.text</span>(<span class="st">&quot;ranked units&quot;</span>, <span class="dt">x =</span> <span class="fl">0.5</span>, <span class="dt">y =</span> <span class="fl">0.05</span>, <span class="dt">default.units =</span> <span class="st">&quot;npc&quot;</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>, <span class="dt">col =</span> <span class="st">&quot;white&quot;</span>))</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-24-1.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="kw">plotHive</span>(EC4, <span class="dt">dr.nodes =</span> <span class="ot">FALSE</span>, <span class="dt">method =</span> <span class="st">&quot;norm&quot;</span>, <span class="dt">ch =</span> <span class="fl">0.1</span>, <span class="dt">axLabs =</span> <span class="kw">c</span>(<span class="st">&quot;source&quot;</span>, <span class="st">&quot;sink&quot;</span>, <span class="st">&quot;manager&quot;</span>),</a>
<a class="sourceLine" id="cb30-2" data-line-number="2"><span class="dt">axLab.pos =</span> <span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>), <span class="dt">axLab.gpar =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">6</span>, <span class="dt">col =</span> <span class="st">&quot;white&quot;</span>), <span class="dt">np =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="kw">grid.text</span>(<span class="st">&quot;normed units&quot;</span>, <span class="dt">x =</span> <span class="fl">0.5</span>, <span class="dt">y =</span> <span class="fl">0.05</span>, <span class="dt">default.units =</span> <span class="st">&quot;npc&quot;</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>, <span class="dt">col =</span> <span class="st">&quot;white&quot;</span>))</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-25-1.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
<div id="load-the-network-data-for-pathways-and-metabolites" class="section level2">
<h2><span class="header-section-number">4.5</span> Load the network data for pathways and metabolites</h2>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1">Gene_stats&lt;-<span class="kw">read.delim</span>(<span class="st">&quot;Circos_HiveR/hiveData/hmrData/GLS_Palm_vs_control_rep_pathway.txt&quot;</span>)</a>
<a class="sourceLine" id="cb31-2" data-line-number="2">met_stats&lt;-<span class="kw">read.delim</span>(<span class="st">&quot;Circos_HiveR/hiveData/hmrData/GSS_Palm_vs_control_rep_met.txt&quot;</span>)</a>
<a class="sourceLine" id="cb31-3" data-line-number="3">path_stats&lt;-<span class="st">  </span><span class="kw">read.delim</span>(<span class="st">&quot;Circos_HiveR/hiveData/hmrData/GSS_Palm_vs_control_rep_pathway.txt&quot;</span>)</a>
<a class="sourceLine" id="cb31-4" data-line-number="4">met_gene&lt;-<span class="kw">read.delim</span>(<span class="st">&quot;Circos_HiveR/hiveData/hmrData/GSC_Palm_vs_control_rep_met.txt&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb31-5" data-line-number="5">path_gene&lt;-<span class="kw">read.delim</span>(<span class="st">&quot;Circos_HiveR/hiveData/hmrData/GSC_Palm_vs_control_rep_pathway.txt&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<p>Take a look at the loaded files.</p>
<p>The code below will create a data of “HivePlotData” class from the output of the Piano package in R. The code at the moment is is not fully commeneted but please have a look at the course homepage for updates. For those who are familiar with R, can try and understand it from the implemented commands.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">met_gene_z&lt;-<span class="st"> </span>met_gene[<span class="kw">which</span>(met_gene<span class="op">$</span>V1 <span class="op">%in%</span><span class="st"> </span>met_stats<span class="op">$</span>Name),]</a>
<a class="sourceLine" id="cb32-2" data-line-number="2">path_gene_z&lt;-path_gene[<span class="kw">which</span>(<span class="kw">as.character</span>(path_gene<span class="op">$</span>V1) <span class="op">%in%</span><span class="st"> </span><span class="kw">as.character</span>(path_stats<span class="op">$</span>Name)),]</a>
<a class="sourceLine" id="cb32-3" data-line-number="3"></a>
<a class="sourceLine" id="cb32-4" data-line-number="4"></a>
<a class="sourceLine" id="cb32-5" data-line-number="5">Gene_stats<span class="op">$</span>ID1&lt;-<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span>(<span class="kw">dim</span>(Gene_stats)[<span class="dv">1</span>]))</a>
<a class="sourceLine" id="cb32-6" data-line-number="6">Gene_stats<span class="op">$</span>padj&lt;-<span class="kw">p.adjust</span>(Gene_stats<span class="op">$</span>p, <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</a>
<a class="sourceLine" id="cb32-7" data-line-number="7">met_gene<span class="op">$</span>ID2&lt;-<span class="kw">as.integer</span>(<span class="kw">as.numeric</span>(<span class="kw">factor</span>(met_gene<span class="op">$</span>V1)))</a>
<a class="sourceLine" id="cb32-8" data-line-number="8">path_gene<span class="op">$</span>ID3&lt;-<span class="kw">as.integer</span>(<span class="kw">as.numeric</span>(<span class="kw">factor</span>(path_gene<span class="op">$</span>V1)))</a>
<a class="sourceLine" id="cb32-9" data-line-number="9"></a>
<a class="sourceLine" id="cb32-10" data-line-number="10">gene_met_tmp&lt;-<span class="kw">merge</span>(met_gene, Gene_stats, <span class="dt">by.x =</span> <span class="st">&quot;V2&quot;</span>, <span class="dt">by.y =</span><span class="st">&quot;g&quot;</span> )</a>
<a class="sourceLine" id="cb32-11" data-line-number="11">gene_met_tmp<span class="op">$</span>ID11&lt;-<span class="kw">as.integer</span>(<span class="kw">as.numeric</span>(<span class="kw">factor</span>(gene_met_tmp<span class="op">$</span>V2)))</a>
<a class="sourceLine" id="cb32-12" data-line-number="12">gene_met_tmp<span class="op">$</span>ID11&lt;-gene_met_tmp<span class="op">$</span>ID11<span class="op">+</span><span class="kw">max</span>(gene_met_tmp<span class="op">$</span>ID2)</a>
<a class="sourceLine" id="cb32-13" data-line-number="13">gene_met_tmp&lt;-<span class="kw">merge</span>(gene_met_tmp, path_gene, <span class="dt">by.x =</span> <span class="st">&quot;V2&quot;</span>, <span class="dt">by.y =</span> <span class="st">&quot;V2&quot;</span> )</a>
<a class="sourceLine" id="cb32-14" data-line-number="14">gene_met_tmp<span class="op">$</span>ID31&lt;-<span class="st"> </span>gene_met_tmp<span class="op">$</span>ID3<span class="op">+</span><span class="kw">max</span>(gene_met_tmp<span class="op">$</span>ID11)</a></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1">met_stats_sig&lt;-(met_stats[met_stats<span class="op">$</span>p.adj..non.dir..<span class="op">&lt;</span><span class="fl">0.05</span>,])</a>
<a class="sourceLine" id="cb33-2" data-line-number="2">path_stats_sig&lt;-path_stats[path_stats<span class="op">$</span>p.adj..non.dir..<span class="op">&lt;</span><span class="fl">0.00001</span>,]</a>
<a class="sourceLine" id="cb33-3" data-line-number="3"></a>
<a class="sourceLine" id="cb33-4" data-line-number="4">sig_mets&lt;-met_stats_sig<span class="op">$</span>Name <span class="op">%&gt;%</span><span class="kw">as.character</span>()</a>
<a class="sourceLine" id="cb33-5" data-line-number="5">sig_paths&lt;-path_stats_sig<span class="op">$</span>Name<span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.character</span>()</a>
<a class="sourceLine" id="cb33-6" data-line-number="6"></a>
<a class="sourceLine" id="cb33-7" data-line-number="7"></a>
<a class="sourceLine" id="cb33-8" data-line-number="8">gene_met_tmp_sig&lt;-gene_met_tmp[<span class="kw">which</span>( <span class="kw">as.character</span>(gene_met_tmp<span class="op">$</span>V1.x)<span class="op">%in%</span><span class="st"> </span>sig_mets<span class="op">|</span><span class="st">  </span><span class="kw">as.character</span>(gene_met_tmp<span class="op">$</span>V1.y)<span class="op">%in%</span><span class="st"> </span>sig_paths),]</a>
<a class="sourceLine" id="cb33-9" data-line-number="9"></a>
<a class="sourceLine" id="cb33-10" data-line-number="10">gene_met_tmp_sig &lt;-gene_met_tmp_sig[gene_met_tmp_sig<span class="op">$</span>padj<span class="op">&lt;</span><span class="fl">0.000001</span>,]</a>
<a class="sourceLine" id="cb33-11" data-line-number="11"></a>
<a class="sourceLine" id="cb33-12" data-line-number="12"><span class="co">#Create a node table</span></a>
<a class="sourceLine" id="cb33-13" data-line-number="13">x_sig&lt;-<span class="kw">unique</span>(<span class="kw">c</span>(gene_met_tmp_sig<span class="op">$</span>ID2, gene_met_tmp_sig<span class="op">$</span>ID11,gene_met_tmp_sig<span class="op">$</span>ID31)) <span class="co">#nodes</span></a>
<a class="sourceLine" id="cb33-14" data-line-number="14">x1_sig&lt;-<span class="kw">as.character</span>((<span class="kw">c</span>(<span class="kw">as.character</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V1.x)), <span class="kw">as.character</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V2)),<span class="kw">as.character</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V1.y))))) <span class="co">#labels</span></a>
<a class="sourceLine" id="cb33-15" data-line-number="15">x2_sig&lt;-<span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="kw">length</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V1.x)), <span class="kw">length</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V2)),<span class="kw">length</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V1.y)))) <span class="co"># assign axis</span></a>
<a class="sourceLine" id="cb33-16" data-line-number="16"></a>
<a class="sourceLine" id="cb33-17" data-line-number="17">x3_sig&lt;-<span class="st">  </span><span class="kw">c</span>(<span class="op">-</span><span class="kw">log10</span>(Gene_stats[<span class="kw">which</span>(x1_sig <span class="op">%in%</span><span class="st"> </span>Gene_stats<span class="op">$</span>g),<span class="st">&quot;padj&quot;</span>])<span class="op">*</span><span class="dv">50</span>, <span class="op">-</span><span class="kw">log10</span>(met_stats[<span class="kw">which</span>(x1_sig <span class="op">%in%</span><span class="st"> </span>met_stats<span class="op">$</span>Name),<span class="st">&quot;p..non.dir..&quot;</span>])<span class="op">*</span><span class="dv">50</span>,<span class="op">-</span><span class="kw">log10</span>(path_stats[<span class="kw">which</span>(path_stats<span class="op">$</span>Name <span class="op">%in%</span><span class="st"> </span>x1_sig), <span class="st">&quot;p..non.dir..&quot;</span>])<span class="op">*</span><span class="dv">100</span> )   <span class="co">#radius i.e., postion on the axis is arranged by meanigful graphical parameter e.g., significance</span></a>
<a class="sourceLine" id="cb33-18" data-line-number="18"></a>
<a class="sourceLine" id="cb33-19" data-line-number="19">x4_sig&lt;-<span class="st">  </span><span class="kw">scale</span>(<span class="kw">abs</span>(<span class="kw">c</span>(Gene_stats[<span class="kw">which</span>(x1_sig <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(Gene_stats)),<span class="st">&quot;myFC&quot;</span>], met_stats_sig[<span class="kw">which</span>(x1_sig <span class="op">%in%</span><span class="st"> </span>met_stats<span class="op">$</span>Name ), <span class="st">&quot;Genes..tot.&quot;</span> ],path_stats[<span class="kw">which</span>(path_stats<span class="op">$</span>Name <span class="op">%in%</span><span class="st"> </span>x1_sig), <span class="st">&quot;Genes..tot.&quot;</span> ] )), <span class="dt">center =</span> <span class="ot">FALSE</span>, <span class="dt">scale =</span> <span class="ot">TRUE</span>) <span class="co">#size of the node e.g., Degree of the node or fold change etc</span></a>
<a class="sourceLine" id="cb33-20" data-line-number="20">x4_sig_a&lt;-<span class="kw">rep</span>(<span class="dv">1</span>, <span class="kw">length</span>(x3_sig))</a>
<a class="sourceLine" id="cb33-21" data-line-number="21"></a>
<a class="sourceLine" id="cb33-22" data-line-number="22">x5_sig&lt;-<span class="st">  </span><span class="kw">as.character</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;#E41A1C&quot;</span> ,<span class="st">&quot;#984EA3&quot;</span>, <span class="st">&quot;#4DAF4A&quot;</span>), <span class="kw">c</span>(<span class="kw">length</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V1.x)), <span class="kw">length</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V2)),<span class="kw">length</span>(<span class="kw">unique</span>(gene_met_tmp_sig<span class="op">$</span>V1.y)))))   <span class="co">#Node color</span></a>
<a class="sourceLine" id="cb33-23" data-line-number="23">  </a>
<a class="sourceLine" id="cb33-24" data-line-number="24"><span class="co">#xx&lt;-c(gene_stats[which(rownames(gene_stats) %in%x1),&quot;myPval&quot;], met_stats[which(met_stats$Name %in% x1), &quot;p..non.dir..&quot;],path_stats[which(path_stats$Name %in% x1), &quot;p..non.dir..&quot;] )</span></a></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="co"># Create and edge table</span></a>
<a class="sourceLine" id="cb34-2" data-line-number="2">y_sig&lt;-<span class="st">  </span><span class="kw">c</span>(gene_met_tmp_sig<span class="op">$</span>ID2, gene_met_tmp_sig<span class="op">$</span>ID11,gene_met_tmp_sig<span class="op">$</span>ID2 )<span class="co">#ID1</span></a>
<a class="sourceLine" id="cb34-3" data-line-number="3">y1_sig&lt;-<span class="st"> </span><span class="kw">c</span>(gene_met_tmp_sig<span class="op">$</span>ID11, gene_met_tmp_sig<span class="op">$</span>ID31,gene_met_tmp_sig<span class="op">$</span>ID31)<span class="co">#ID2</span></a>
<a class="sourceLine" id="cb34-4" data-line-number="4">y2_sig&lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">2</span>, <span class="kw">length</span>(y_sig)) <span class="co"># weight</span></a>
<a class="sourceLine" id="cb34-5" data-line-number="5">y3_sig&lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">rep</span> (<span class="st">&quot;pink&quot;</span> , <span class="kw">length</span>(y_sig)))<span class="co">#color</span></a></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1">HivePlotData&lt;-<span class="kw">list</span>()</a>
<a class="sourceLine" id="cb35-2" data-line-number="2">nodes_sig&lt;-<span class="kw">data.frame</span>(x_sig,x1_sig,x2_sig,x3_sig,x4_sig_a,x5_sig)</a>
<a class="sourceLine" id="cb35-3" data-line-number="3"><span class="kw">colnames</span>(nodes_sig)&lt;-<span class="kw">c</span>(<span class="st">&quot;id&quot;</span>, <span class="st">&quot;lab&quot;</span>, <span class="st">&quot;axis&quot;</span>,<span class="st">&quot;radius&quot;</span>, <span class="st">&quot;size&quot;</span>, <span class="st">&quot;color&quot;</span>)</a>
<a class="sourceLine" id="cb35-4" data-line-number="4">edges_sig&lt;-<span class="st"> </span><span class="kw">data.frame</span>(y_sig, y1_sig, y2_sig, y3_sig)</a>
<a class="sourceLine" id="cb35-5" data-line-number="5"><span class="kw">colnames</span>(edges_sig)&lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;id1&quot;</span>, <span class="st">&quot;id2&quot;</span>, <span class="st">&quot;weight&quot;</span>, <span class="st">&quot;color&quot;</span>)</a>
<a class="sourceLine" id="cb35-6" data-line-number="6">type&lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;2D&quot;</span>)</a>
<a class="sourceLine" id="cb35-7" data-line-number="7">desc&lt;-<span class="kw">c</span>(<span class="st">&quot;many nodes&quot;</span>)</a>
<a class="sourceLine" id="cb35-8" data-line-number="8"><span class="co">#axis.cols&lt;- c(&quot;#E41A1C&quot; ,&quot;#377EB8&quot;, &quot;#4DAF4A&quot;)</span></a>
<a class="sourceLine" id="cb35-9" data-line-number="9">axis.cols&lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;blue&quot;</span> ,<span class="st">&quot;purple&quot;</span>, <span class="st">&quot;green&quot;</span>)</a>
<a class="sourceLine" id="cb35-10" data-line-number="10"></a>
<a class="sourceLine" id="cb35-11" data-line-number="11">Data_hive_sig&lt;-<span class="st"> </span><span class="kw">list</span>(nodes_sig, edges_sig, type, desc, axis.cols)</a>
<a class="sourceLine" id="cb35-12" data-line-number="12"></a>
<a class="sourceLine" id="cb35-13" data-line-number="13"><span class="kw">names</span>(Data_hive_sig)&lt;-<span class="kw">c</span>(<span class="st">&quot;nodes&quot;</span>, <span class="st">&quot;edges&quot;</span>, <span class="st">&quot;type&quot;</span>, <span class="st">&quot;desc&quot;</span>, <span class="st">&quot;axis.cols&quot;</span>)</a>
<a class="sourceLine" id="cb35-14" data-line-number="14">Data_hive_sig<span class="op">$</span>nodes<span class="op">$</span>lab&lt;-<span class="kw">as.character</span>(Data_hive_sig<span class="op">$</span>nodes<span class="op">$</span>lab)</a>
<a class="sourceLine" id="cb35-15" data-line-number="15">Data_hive_sig<span class="op">$</span>nodes<span class="op">$</span>color&lt;-<span class="kw">as.character</span>(Data_hive_sig<span class="op">$</span>nodes<span class="op">$</span>color)</a>
<a class="sourceLine" id="cb35-16" data-line-number="16">Data_hive_sig<span class="op">$</span>edges<span class="op">$</span>color&lt;-<span class="kw">as.character</span>(Data_hive_sig<span class="op">$</span>edges<span class="op">$</span>color)</a>
<a class="sourceLine" id="cb35-17" data-line-number="17"></a>
<a class="sourceLine" id="cb35-18" data-line-number="18"></a>
<a class="sourceLine" id="cb35-19" data-line-number="19"></a>
<a class="sourceLine" id="cb35-20" data-line-number="20"><span class="kw">class</span>(Data_hive_sig)&lt;-<span class="st">&quot;HivePlotData&quot;</span></a></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw">plotHive</span>(Data_hive_sig, <span class="dt">bkgnd =</span> <span class="st">&quot;black&quot;</span>,</a>
<a class="sourceLine" id="cb36-2" data-line-number="2">    <span class="dt">axLabs =</span> <span class="kw">c</span>(<span class="st">&quot;Metabolites&quot;</span>, <span class="st">&quot;Genes&quot;</span>, <span class="st">&quot;Pathways&quot;</span>),</a>
<a class="sourceLine" id="cb36-3" data-line-number="3">    <span class="dt">axLab.pos =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>),</a>
<a class="sourceLine" id="cb36-4" data-line-number="4">    <span class="dt">ch =</span> <span class="fl">0.5</span>, <span class="dt">method =</span> <span class="st">&quot;norm&quot;</span></a>
<a class="sourceLine" id="cb36-5" data-line-number="5">    )</a></code></pre></div>
<img src="RCircos-HiveR_files/figure-html/unnamed-chunk-31-1.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
